download_images.py

```diff
@@ -0,0 +1,71 @@
+import concurrent.futures
+import itertools
+import json
+import pandas as pd
+import socket
+import urllib.request
+import os
+from os.path import join, exists
+
+class ImageDownloader(object):
+
+    def __init__(self, image_url_df, target_path='data/images', nb_workers=64):
+        self._failed_to_capture_path = 'work/failed_to_capture_images.json'
+        self.image_urls = image_url_df.itertuples()
+        self.target_path = target_path
+        self.nb_workers = nb_workers
+        self.imgs_to_request = image_url_df.shape[0]
+        self.imgs_requested = 0
+        self.failed_img_requests = 0
+        socket.setdefaulttimeout(5)
+
+    def __enter__(self):
+        try:
+            if not exists(self.target_path):
+                os.makedirs(self.target_path)
+            if not exists('work'):
+                os.makedirs('work')
+            with open(self._failed_to_capture_path, 'r') as f:
+                self.failed_to_capture = json.load(f)
+        except:
+            self.failed_to_capture = []
+        return(self)
+
+    def __exit__(self, *args):
+        with open(self._failed_to_capture_path, 'w') as f:
+            json.dump(self.failed_to_capture, f)
+
+    def run(self):
+        print('Images to request: ', self.imgs_to_request)
+        with concurrent.futures.ThreadPoolExecutor(max_workers=self.nb_workers) as worker_pool:
+            for file_num, fname, img_src in self.image_urls:
```

Comment:
Is the `file_num` necessary/useful here? I think maybe if `enumerate` had been used, it might be. But, it doesn't look like it's used. Was it meant to be?